plugins/overwatch.rb

```diff
@@ -0,0 +1,660 @@
+
+          ip_addr = interface.addrs[index]
+          netmask = interface.netmasks[index]
+
+          next unless ip_addr =~ /\./ # Pick out the IPv4 addresses
+          subnet = get_subnet(ip_addr, netmask)
+          next unless is_routable?(subnet, netmask)
+
+          if subnet
+            remove_if_stale(subnet, netmask, sid, stale_sids) if self.config[:reroute_stale]
+
+            if !switch_board.route_exists?(subnet, netmask)
+              begin
+                if Rex::Socket::SwitchBoard.add_route(subnet, netmask, framework.sessions[sid])
+                  session_log(sid, "Route added to subnet #{subnet}/#{netmask} from #{interface.mac_name}.")
+                else
+                  session_log(sid, "Could not add route to subnet #{subnet}/#{netmask} from #{interface.mac_name}")
+                end
+              rescue ::Exception => error
+                session_log(sid, "Could not add route to subnet #{subnet}/#{netmask} from #{interface.mac_name}")
+                session_log(sid, error.to_s)
+              end
+            end
+          end 
+        end
+      end
+    end
+
+    # Removes routes associated with stale sessions only if they are
+    # also present in sessions that are still active.
+    #
+    # @subnet [string class] route subnet
+    # @netmask [string array] route netmask
+    # @curr_sid [int class] Session ID of the current session being processed
+    # @stale_sids [int array] Array of session ID's that have been marked as stale.
+    #
+    # @return [void] A useful return value is not expected here
+    def remove_if_stale(subnet, netmask, curr_sid, stale_sids)
+      Rex::Socket::SwitchBoard.each do | route |
+        if stale_sids.include? route.comm.sid # See if a route is associated with a stale session
+          if route.subnet == subnet && route.netmask == netmask # See if that route matches the one currently being processed
+            session_log(route.comm.sid, "Deleting route #{route.subnet}/#{route.netmask}")
+            begin
+              Rex::Socket::SwitchBoard.remove_route(route.subnet, route.netmask, route.comm) # Remove so fresh matching route can be added
+            rescue ::Exception => error
+              session_log(route.comm.sid, "Could not remove route to subnet #{route.subnet}/#{route.netmask}")
+              session_log(route.comm.sid, error.to_s)
+            end
+            return
+          end
+        end
+      end
+    end
+
+    # Take an IP address and a netmask and return the appropreate subnet "Network"
+    #
+    # @ip_addr [string class] Input IPv4 Address
+    # @netmask [string class] Input IPv4 Netmask
+    #
+    # @return [string class] The subnet related to the IP address and netmask
+    # @return [nil class] Something is out of range
+    def get_subnet(ip_addr, netmask)
+      return nil if !validate_cmd(ip_addr, netmask) #make sure IP and netmask are valid
+
+      nets = ip_addr.split('.')
```

Comment:
I wanted to get this out with IPv4 first and then look into adding IPv6 routing. Adding v6 routes from the hosts' routing tables shouldn't be too big of an issue other than some checks. However, adding them from interfaces will be more challenging. I need some time to review/research on IPv6 since I haven't done much with it in a couple of years. Will update the PR to note that v6 is not yet supported in this plugin, but that future work on it is planned.